presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchQueryBuilder.java

```diff
@@ -0,0 +1,254 @@
+                }
+            });
+        }
+        return elasticQueryBuilder.hasClauses() ? elasticQueryBuilder : new MatchAllQueryBuilder();
+    }
+
+    private QueryBuilder buildPredicate(String columnName, Domain domain, Type type, boolean isList)
+    {
+        checkArgument(domain.getType().isOrderable(), "Domain type must be orderable");
+        BoolQueryBuilder elasticQueryBuilder = new BoolQueryBuilder();
+
+        if (domain.getValues().isNone() && domain.isNullAllowed()) {
+            elasticQueryBuilder.mustNot(new ExistsQueryBuilder(columnName));
+            return elasticQueryBuilder;
+        }
+
+        if (domain.getValues().isAll()) {
+            elasticQueryBuilder.must(new ExistsQueryBuilder(columnName));
+            return elasticQueryBuilder;
+        }
+
+        if (isList) {
+            return buildMatchQuery(elasticQueryBuilder, columnName, domain, type);
+        }
+        return buildTermQuery(elasticQueryBuilder, columnName, domain, type);
+    }
+
+    private QueryBuilder buildMatchQuery(BoolQueryBuilder builder, String columnName, Domain domain, Type type)
+    {
+        List<Object> includeValues = new ArrayList<>();
+        List<Object> excludeValues = new ArrayList<>();
+        for (Range range : domain.getValues().getRanges().getOrderedRanges()) {
+            checkState(!range.isAll());
+            if (range.isSingleValue()) {
+                includeValues.add(range.getLow().getValue());
+            }
+
+            if (!range.getLow().isLowerUnbounded() && range.getLow().getBound() == ABOVE) {
+                Object value = range.getLow().getValue();
+                if (!excludeValues.contains(value)) {
+                    excludeValues.add(value);
+                }
+            }
+
+            if (!range.getHigh().isUpperUnbounded() && range.getHigh().getBound() == BELOW) {
+                Object value = range.getHigh().getValue();
+                if (!excludeValues.contains(value)) {
+                    excludeValues.add(value);
+                }
+            }
+        }
+
+        for (Object value : excludeValues) {
+            builder.mustNot(new MatchQueryBuilder(columnName, getValue(type, value)));
+        }
+
+        if (includeValues.size() == 1) {
+            builder.must(new MatchQueryBuilder(columnName, getValue(type, getOnlyElement(includeValues))));
+        }
+        else {
+            for (Object value : includeValues) {
+                builder.should(new MatchQueryBuilder(columnName, getValue(type, value)));
+            }
+        }
+        return builder;
+    }
+
+    private QueryBuilder buildTermQuery(BoolQueryBuilder builder, String columnName, Domain domain, Type type)
+    {
+        List<Object> values = new ArrayList<>();
+        for (Range range : domain.getValues().getRanges().getOrderedRanges()) {
+            checkState(!range.isAll());
+            if (range.isSingleValue()) {
+                values.add(range.getLow().getValue());
```

Comment:
you can add a `continue` after L196 and get rid of the big `else` block surrounding the two big if/switches below (so that we get rid of some nesting here).

Reply:
you can also extract two separate methods for each switch statement below to simplify this method further.